### PR TITLE
[Intel][XPU] Add pyxccl: direct Intel oneCCL communicator for XPU

### PR DESCRIPTION
**Pull request by siju-samuel**
## Motivation

On Intel XPU, tensor-parallel (`--tp > 1`) collectives currently dispatch through 
`torch.distributed`'s XCCL backend. This is useful when we want to do some quick 
prototype on the latest oneccl releases without the full torch/xccl package. Some 
torch XPU builds may ship **without a usable XCCL backend** , which leaves 
tensor parallelism with no working collective path.

This PR adds `pyxccl` — the XPU counterpart of the existing CUDA `pynccl` — a
direct binding to Intel oneCCL's NCCL-compatible C API (`libccl.so`) that
bypasses `torch.distributed`. It gives XPU an in-tree collective path that does
not depend on torch shipping XCCL, mirroring how CUDA uses `pynccl` to call
NCCL directly.

## Modifications

**New**
- `python/sglang/srt/distributed/device_communicators/pyxccl_wrapper.py` —
  vendored `ctypes` binding of oneCCL's `oneccl*` C API from `libccl.so` (the
  XPU analogue of `pynccl_wrapper.py`). No third-party Python package is
  required; only the oneCCL runtime library.
- `python/sglang/srt/distributed/device_communicators/pyxccl.py` —
  `PyXcclCommunicator` (all_reduce / all_gather / reduce_scatter / broadcast /
  send / recv / destroy), analogous to `PyNcclCommunicator`. Uses the XPU SYCL
  queue (`torch.xpu.current_stream().sycl_queue`) as the oneCCL stream handle.
- `test/registered/xpu/test_pyxccl_collectives.py` — 2-XPU numeric tests for all
  four collectives (× float32/bfloat16/float16) plus an end-to-end SGLang
  TP-collective integration test.
- `docs_new/docs/hardware-platforms/xpu.pyxccl.md` — setup guide.
- `scripts/build_oneccl_xpu.sh`, `scripts/activate-xpu-sglang.sh` — SYCL oneCCL
  build + runtime activation helpers.

**Wiring** (`parallel_state.py`)
- `pyxccl_comm` added as a sibling of `pynccl_comm` on `GroupCoordinator`,
  branched **before** the `torch.distributed` fallback in `_all_reduce_in_place`,
  `_reduce_scatter_tensor`, `_all_gather_into_tensor`, `broadcast`, `send`,
  `recv`, and released in `destroy()`.
- Only constructed for XPU model-parallel groups, so CUDA / ROCm / NPU / CPU
  paths are byte-for-byte unchanged.

**Env vars** (`environ.py`)
- `SGLANG_ENABLE_PYXCCL` (default `False`) — opt in to the direct-oneCCL path.
- `SGLANG_PYXCCL_SO_PATH` — explicit path to `libccl.so.1` (else the dynamic
  linker resolves it).
- `SGLANG_DEBUG_PYXCCL` — log every collective call to confirm the oneCCL path.

**Behavior**
- Off by default; XPU `tp>1` uses `torch.distributed` / XCCL exactly as before.
- With `SGLANG_ENABLE_PYXCCL=1`, if oneCCL cannot initialize at `tp>1`, SGLang
  raises at startup with setup guidance rather than silently falling back
  (the opt-in usually means XCCL is unavailable, so a silent fallback would be
  wrong).

## Accuracy Tests

Verified on 2× Intel Arc Pro B60 (oneCCL 2021.17.2), `torch 2.12.0+xpu`:

- `test/registered/xpu/test_pyxccl_collectives.py` — **5 passed**:
  `all_reduce`, `all_gather`, `broadcast`, `reduce_scatter` (numerically checked
  against closed-form expected values across float32/bfloat16/float16 at tp=2),
  plus `test_sglang_tp_collectives` (public TP API →
  `initialize_model_parallel` → `GroupCoordinator` → pyxccl).
- End-to-end `bench_one_batch` (Qwen2.5-1.5B-Instruct, `--device xpu --tp 2`):
  - `SGLANG_ENABLE_PYXCCL=1` → generation completes; `SGLANG_DEBUG_PYXCCL=1`
    shows collectives routed through oneCCL (930 calls: 914 AllReduce +
    16 AllGather).
  - flag unset → same run completes via `torch.distributed` (0 pyxccl calls),
    confirming the fallback path is unaffected.

## Speed Tests and Profiling

Not a perf-oriented change (adds an alternative communicator; default path
unchanged). Functional throughput sanity on Qwen2.5-1.5B tp=2 was comparable
between the pyxccl and torch.distributed paths; no rigorous benchmarking was
performed.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: docs_new/docs/hardware-platforms/xpu.pyxccl.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+`python/sglang/srt/distributed/device_communicators/pyxccl.py`. The ctypes layer
+that binds oneCCL's `oneccl*` C API is vendored in
+`pyxccl_wrapper.py` (the XPU analogue of `pynccl_wrapper.py`) — no third-party
+Python binding is required, only the oneCCL runtime library `libccl.so`.
+
+By default, tensor-parallel (`--tp > 1`) collectives on Intel GPU go through
+torch's XCCL backend. **pyxccl is opt-in**: set `SGLANG_ENABLE_PYXCCL=1` to route
+them through oneCCL directly instead. This is the path to use when your torch
+build has no usable XCCL backend. When `SGLANG_ENABLE_PYXCCL=1` is set but pyxccl
+cannot initialize, SGLang raises at startup for `tp>1` rather than silently
```
Comment (polisettyvarma) [inline docs_new/docs/hardware-platforms/xpu.pyxccl.md:16]: tp=1 flow also triggers distributed init, do we need to change anything here ?

File: docs_new/docs/hardware-platforms/xpu.pyxccl.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+---
+
+## Quick Checklist
+
+```text
+[ ] 0. Install one-time system prerequisites
+[ ] 1. Source Intel oneAPI compiler environment
+[ ] 2. Build oneCCL-v2 XPU library (libccl.so.1)
+[ ] 3. Point SGLANG_PYXCCL_SO_PATH / LD_LIBRARY_PATH at libccl.so.1
+[ ] 4. Source the SGLang-XPU activation script before every run
```
Comment (polisettyvarma) [inline docs_new/docs/hardware-platforms/xpu.pyxccl.md:36]: every run or once per terminal session ?

File: docs_new/docs/hardware-platforms/xpu.pyxccl.md [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+uv venv /work/.build-env --python 3.12
+uv pip install --python /work/.build-env/bin/python cmake ninja
+```
+
+---
+
+## Step 1 — Source the Intel oneAPI compiler
+
+```bash
+source /work/compiler/setvars.sh
```
Comment (polisettyvarma) [inline docs_new/docs/hardware-platforms/xpu.pyxccl.md:62]: assumes oneapi installation is done at /work due to guide instead add a note as well ?

File: docs_new/docs/hardware-platforms/xpu.pyxccl.md [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+lib = ONECCLLibrary('$CCL_INSTALL/lib/libccl.so.1')
+print('oneCCL version:', lib.onecclGetVersion())
+"
+```
+
+---
+
+## Step 4 — Activation script
+
+Source `scripts/activate-xpu-sglang.sh` before every SGLang run. Its contents:
```
Comment (polisettyvarma) [inline docs_new/docs/hardware-platforms/xpu.pyxccl.md:165]: same as above

File: docs_new/docs/hardware-platforms/xpu.pyxccl.md [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+# 3. oneCCL / OFI transport
+export FI_PROVIDER_PATH=${_CCL}/opt/mpi/libfabric/lib/prov
+export CCL_ATL_TRANSPORT=ofi
+export CCL_TOPO_FABRIC_VERTEX_CONNECTION_CHECK=0
+
+# 4. SGLang pyxccl integration
+# Opt into the direct-oneCCL path (pyxccl is off by default).
+export SGLANG_ENABLE_PYXCCL=1
+export SGLANG_PYXCCL_SO_PATH=${_CCL}/lib/libccl.so.1
```
Comment (polisettyvarma) [inline docs_new/docs/hardware-platforms/xpu.pyxccl.md:191]: this path and expected output of lib mismatch

File: python/sglang/srt/distributed/device_communicators/pyxccl.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    Mirrors the interface of :class:`PyNcclCommunicator` so that
+    ``GroupCoordinator`` can use it as a drop-in replacement for the
+    ``torch.distributed`` XPU collective calls. It is opt-in
+    (``SGLANG_ENABLE_PYXCCL=1``) and intended for environments where the torch
+    build has no usable XCCL backend.
+
+    Unlike ``PyNcclCommunicator`` (which starts ``disabled`` and is only enabled
+    inside CUDA-graph capture via ``change_state``), pyxccl is enabled as soon as
+    it initializes: it is the live collective path in both eager and graph modes
+    on XPU, replacing the ``torch.distributed`` calls.
```
Comment (polisettyvarma) [inline python/sglang/srt/distributed/device_communicators/pyxccl.py:52]: why do we need to deviate ?

File: python/sglang/srt/distributed/device_communicators/pyxccl_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+    onecclHalf = 6
+    onecclFloat32 = 7
+    onecclFloat = 7
+    onecclFloat64 = 8
+    onecclDouble = 8
+    onecclBfloat16 = 9
+
+    @classmethod
+    def from_torch(cls, dtype: torch.dtype) -> int:
+        if dtype == torch.int8:
```
Comment (polisettyvarma) [inline python/sglang/srt/distributed/device_communicators/pyxccl_wrapper.py:90]: dict lookup better ?

File: python/sglang/srt/distributed/device_communicators/pyxccl_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+    # Values from onecclRedOp_t in ccl/v2/types.h.
+    onecclSum = 0
+    onecclProd = 1
+    onecclMax = 2
+    onecclMin = 3
+    onecclAvg = 4
+
+    @classmethod
+    def from_torch(cls, op: ReduceOp) -> int:
+        if op == ReduceOp.SUM:
```
Comment (polisettyvarma) [inline python/sglang/srt/distributed/device_communicators/pyxccl_wrapper.py:122]: dict map ?

File: test/registered/xpu/test_pyxccl_collectives.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+from sglang.test.ci.ci_register import register_xpu_ci
+from sglang.test.test_utils import CustomTestCase, find_available_port
+
+register_xpu_ci(est_time=120, suite="stage-a-test-2-gpu-xpu")
+
+WORLD_SIZE = 2
```
Comment (polisettyvarma) [inline test/registered/xpu/test_pyxccl_collectives.py:35]: can we query the device count ?

File: test/registered/xpu/test_pyxccl_collectives.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+# oneCCL supports these element types via onecclDataTypeEnum.from_torch.
+DTYPES = [torch.float32, torch.bfloat16, torch.float16]
```
Comment (polisettyvarma) [inline test/registered/xpu/test_pyxccl_collectives.py:37]: we have lower bit data types also, can we add them ?

File: python/sglang/srt/distributed/device_communicators/pyxccl.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+        self.oneccl.onecclRecv(
+            buffer_type(tensor.data_ptr()),
+            tensor.numel(),
+            onecclDataTypeEnum.from_torch(tensor.dtype),
+            src,
+            self.comm,
+            self._stream(),
+        )
+
+    def destroy(self):
```
Comment (msinnha1) [inline python/sglang/srt/distributed/device_communicators/pyxccl.py:333]: There is no "destroy" comm in PyNcclCommunicator, the primary reason could be to manage it from python across all ranks. Are we using it differently to manage it?

File: python/sglang/srt/distributed/device_communicators/pyxccl.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+            return
+        assert input_tensor.device == self.device, (
+            f"this pyxccl communicator is created to work on {self.device}, "
+            f"but the input tensor is on {input_tensor.device}"
+        )
+        assert input_tensor.numel() % self.world_size == 0, (
+            f"all_to_all_single requires an even split, but numel "
+            f"({input_tensor.numel()}) is not divisible by world_size "
+            f"({self.world_size})"
+        )
```
Comment (msinnha1) [inline python/sglang/srt/distributed/device_communicators/pyxccl.py:247]: missing output checks as present in pynccl.py, ref: https://github.com/sgl-project/sglang/blob/407a65d3cb5eb38873b34de75d3bcd76be5a0bee/python/sglang/srt/distributed/device_communicators/pynccl.py#L339
